numba/np/arrayobj.py

```diff
@@ -5439,3 +5439,92 @@ def impl(arr, axis1, axis2):
+
+
+@register_jitable
+def _take_along_axis_impl_set_ni_nk(arr, indices, axis, Ni, Nk):
+    for i in range(len(Ni)):
+        Ni = tuple_setitem(Ni, i, arr.shape[i])
+    for i in range(len(Nk)):
+        Nk = tuple_setitem(Nk, i, arr.shape[axis + 1 + i])
+    return _take_along_axis_impl(arr, indices, axis, Ni, Nk)
+
+
+@register_jitable
+def _take_along_axis_impl(arr, indices, axis, Ni, Nk):
+    # Based on example code in
+    # https://numpy.org/doc/stable/reference/generated/numpy.take_along_axis.html
+    J = indices.shape[axis]  # Need not equal M
+    out = np.empty(Ni + (J,) + Nk, arr.dtype)
+
+    for ii in np.ndindex(Ni):
+        for ii in np.ndindex(Ni):
+            for kk in np.ndindex(Nk):
+                np_s_ = (slice(None, None, None),)
+                a_1d = arr[ii + np_s_ + kk]
+                indices_1d = indices[ii + np_s_ + kk]
+                out_1d = out[ii + np_s_ + kk]
+                for j in range(J):
+                    out_1d[j] = a_1d[indices_1d[j]]
+    return out
+
+
+@overload(np.take_along_axis, jit_options={'boundscheck':True})
+def arr_take_along_axis(arr, indices, axis):
+    if is_nonelike(axis):
+        def take_along_axis_impl(arr, indices, axis):
+            return _take_along_axis_impl(arr.flatten(), indices, 0, (), ())
+    else:
+        check_is_integer(axis, "axis")
+        if not isinstance(axis, types.IntegerLiteral):
+            raise ValueError(
+                "axis must be a literal value (i.e. not an argument) for now"
+            )
+        axis = axis.literal_value
+        if axis < 0:
+            axis = arr.ndim + axis
+
+        if axis < 0 or axis >= arr.ndim:
+            raise ValueError("axis is out of bounds")
+
+        Ni = tuple(range(axis))
+        Nk = tuple(range(axis + 1, arr.ndim))
+        # We need to do this rather verbose approach because tuple_setitem on
+        # an empty tuple will fail with a LoweringError. If the code were to
+        # _run_, it would be fine, because tuple_setitem isn't actually ever
+        # called, but we never reach the point where the code can be run.
+        if Ni:
+            if Nk:
+                def take_along_axis_impl(arr, indices, axis):
+                    return _take_along_axis_impl_set_ni_nk(
+                        arr, indices, axis, Ni, Nk
+                    )
+            else:
+                def take_along_axis_impl(arr, indices, axis):
+                    return _take_along_axis_impl_set_ni(
+                        arr, indices, axis, Ni, Nk
+                    )
+        else:
+            if Nk:
+                def take_along_axis_impl(arr, indices, axis):
+                    return _take_along_axis_impl_set_nk(
+                        arr, indices, axis, Ni, Nk
+                    )
+            else:
+                def take_along_axis_impl(arr, indices, axis):
+                    return _take_along_axis_impl(arr, indices, axis, Ni, Nk)
+    return take_along_axis_impl
```

Comment:
I think the above can be collapsed into something like the following: https://github.com/stuartarchibald/numba/commit/44451e3c7c72299c241c18cf8cf8ae8a2152aa41

Numba can spot quite a few common patterns like "is argument X None" or "Is the length of this tuple > const" and specialise the code based on this such that it can survive type inference. Have a look and see what you think, you are welcome to incorporate the patch/ideas.

Reply:
I can try, but my experience is that Numba gets very angry at any use of tuple_setitem where the tuple length is zero, _even_ if the logic is such that it will never be called. Possibly an `if len(t) > 0` will work better than a `for i in range(len(t))` though, if it's catching only specific hardcoded patterns.

Reply:
Dead loop elimination is https://github.com/numba/numba/issues/7245, which could eventually handle removal of known dead loops. The already existing dead branch pruning will pick up what it can infer from types and eliminate anything it can concretely work out as dead. In general, type inference will assume that all code is reachable and will attempt to propagate types, it has to do this static analysis so as to create a concretely typed specialisation that it can compile.


The code in the patch above is a translation of what's in this PR into something that makes use of more of what the compiler can work out so as to shrink the code.

Reply:
I did more or less the equivalent in my PR. Couldn't quite use it because in the axis=None case the array has to be flattened.